Feature(UI): Add text tool to canvas

DustyShoe opened the pull request:
## Summary

Canvas Basic Text Tool!


Summary:

* Adds a basic Text tool with overlay editing and rasterize-on-commit.
* New text options row (font, size, bold/italic/underline/strike, alignment).
* Handles hotkey suppression during typing, commit on Enter/tool switch.
* Overlay positioning, tests for text helpers.




https://github.com/user-attachments/assets/70c94abe-ad83-4de8-bedb-b6d3143fc5f5




## Related Issues / Discussions

The fonts information is stored in `invokeai\frontend\web\src\features\controlLayers\text\textConstants.ts`
If anyone have an opinion what fonts and fallbacks should be used, please write in comments. There's 10 fonts to choose ATM and i'd like to make selection as diverse as possible.

## QA Instructions

* Activate Text tool, click canvas, type 2–3 lines, Shift+Enter adds newline, Enter commits. Esc cancels operation.
* Verify that typing works as expected, hotkeys are suppressed while typing (only Ctrl/Cmd+C/V/Z/Y allowed).
* Change font, size, bold/italic/underline/strike, alignment; preview matches raster after commit.
* Click outside text box or switch tools; commit happens once, creates a raster layer above selected layer.

## Merge Plan

<!--WHEN APPLICABLE: Large PRs, or PRs that touch sensitive things like DB schemas, may need some care when merging. For example, a careful rebase by the change author, timing to not interfere with a pending release, or a message to contributors on discord after merging.-->

## Checklist

- [x] _The PR has a short but descriptive title, suitable for a changelog_
- [ ] _Tests added / updated (if applicable)_
- [ ] _❗Changes to a redux slice have a corresponding migration_
- [ ] _Documentation added / updated (if applicable)_
- [ ] _Updated `What's New` copy (if doing a release after this PR)_

DustyShoe committed:
Initial mashup of mentioned feature. Still need to resolve some quirks and kinks.

DustyShoe committed:
Merge branch 'invoke-ai:main' into Feature/Add-Text-tool-to-canvas

DustyShoe committed:
Clean text tool integration

DustyShoe committed:
Merge branch 'Feature/Add-Text-tool-to-canvas' of https://github.com/DustyShoe/InvokeAI into Feature/UI-Add-Text-tool-to-canvas

DustyShoe committed:
Merge branch 'invoke-ai:main' into Feature/Add-Text-tool-to-canvas

DustyShoe committed:
Merge branch 'invoke-ai:main' into Feature/Add-Text-tool-to-canvas

DustyShoe committed:
Fixed text tool opions bar jumping and added more fonts

DustyShoe committed:
Touch up for cursor styling

DustyShoe committed:
Minor addition to doc file

DustyShoe committed:
Appeasing frontend checks

DustyShoe committed:
Prettier fix

DustyShoe committed:
knip fixes

DustyShoe committed:
Added safe zones to font selection and color picker to be clickable without commiting text.

DustyShoe committed:
Removed color probing on cursor and added dynamic font display for fallback, minor tweaks

DustyShoe committed:
Finally fixed the text shifting on commit

DustyShoe committed:
Cursor now represent actual input field size. Tidy up options UI

DustyShoe committed:
Some strikethrough and underline line tweaks

DustyShoe commented on invokeai/frontend/web/src/features/controlLayers/components/Text/CanvasTextOverlay.tsx:
```diff
@@ -0,0 +1,337 @@
+
+    let attempts = 0;
+    const MAX_ATTEMPTS = 5;
+    const tryFocus = () => {
```
What would be the better solution here in your opinion?

DustyShoe committed:
Replaced the focus retry loop with a callback‑ref based approach in in CanvasTextOverlay.tsx
Renamed containerMetrics to textContainerData in CanvasTextOverlay.tsx
Fixed mouse cursor disapearing during typing.

DustyShoe committed:
Merge branch 'main' into Feature/Add-Text-tool-to-canvas

DustyShoe committed:
Added missing localistaion string

DustyShoe committed:
Merge branch 'main' into Feature/Add-Text-tool-to-canvas

DustyShoe committed:
Merge branch 'invoke-ai:main' into Feature/Add-Text-tool-to-canvas

ufuksarp commented:
Is it possible to pick from the fonts that are installed on the system?
Also tracking setting seems essential.

DustyShoe commented:
> Is it possible to pick from the fonts that are installed on the system?

Unfortunately not. Because of security and privacy restrictions.
Web applications run in a sandbox and are not allowed to freely access system resources, including installed fonts, since this could be used for fingerprinting users or leaking system information.

As a workaround i was planning to add folder where you can place your fonts and app will scan for them. But that requires digging into backend and i want to avoid it at that stage.

> Also tracking setting seems essential.

 Not sure what you mean.

ufuksarp commented:
That’s unfortunate.

Tracking is the spacing between the characters.

DustyShoe committed:
Merge branch 'main' into Feature/Add-Text-tool-to-canvas

DustyShoe commented:
Wouldn't it work against the claim, that Invoke can be operated completely offline?

DustyShoe committed:
Merge branch 'main' into Feature/Add-Text-tool-to-canvas

lstein committed:
Merge branch 'main' into Feature/Add-Text-tool-to-canvas

lstein commented on docs/development/canvas-text-tool.md:
```diff
@@ -0,0 +1,35 @@
+# Canvas Text Tool
```
It makes perfect sense to put this into `docs/development`, however developer's documentation for the frontend has been going into `docs/contributing/frontend` and I'd prefer to keep it all in one place. (We may reorganize the developer's documentation in the future.) Could you rename this file?

lstein reviewed (changes requested):
I've made a request about where to put the documentation file. I've tested the feature and it works as advertised, but I'm not strong enough with typescript to comment on the code.

DustyShoe committed:
Moved canvas-text-tool.md to docs/contributing/frontend

blessedcoolant committed:
ui: Improve functionality of the text toolbar

Few things done with this commit.

- The varying size of the font selector box has been fixed. The UI no longer shifts and moves with font change.
- We no longer format the font size input to add px each time. Instead now just have a permanent px indicator.
- The bug with the random text inputs on the slider value has also been fixed.
- The font size value is only committed on blur keeping it consistent with other editing apps.
- Fixed the spacing of the toolbar to make it look cleaner.
- Font size now permits increments of 1.

DustyShoe committed:
Added autoselect text in font size on click allowing immediate imput

DustyShoe committed:
Improvement: Added uncommited layer state with CTRL-move and options to select line spacing.

DustyShoe committed:
Added rotation handle to rotate uncommiitted text layer.

DustyShoe committed:
Fix: Redirect user facing labels to use localization file + Add tool discription to docs

DustyShoe committed:
Fixed box padding. Disable tool swich when text input is active, added message on canvas for better UX.

DustyShoe committed:
Updated  Text tool description

DustyShoe committed:
Updated  Text tool description

DustyShoe committed:
Merge branch 'Feature/Add-Text-tool-to-canvas' of https://github.com/DustyShoe/InvokeAI into Feature/Add-Text-tool-to-canvas

DustyShoe commented:
I'd like to ask to take another look at this PR.
Have made some changes to text layer behaviour. It is now movable with CTRL and can be rotated. Also fixed ugly bottom padding. 
We commit only on `Enter` allowing freely click around UI. Switching to other tools is blocked while text input is active. Changing tabs (Generate, Upsacaling etc.) discards uncommitted text.

DustyShoe committed:
Typo

DustyShoe committed:
Add draggable text-box border with improved cursor feedback and larger hit targets. Supress hotkeys on uncommitted text.

DustyShoe committed:
Merge branch 'main' into Feature/Add-Text-tool-to-canvas

DustyShoe committed:
Lint

DustyShoe committed:
Merge branch 'invoke-ai:main' into Feature/Add-Text-tool-to-canvas

DustyShoe committed:
Fix(bug): text commit to link uploaded image assets instead of embedding full base64

lstein reviewed (approved):
I think we're good to merge this PR in and wait to see what feedback we get from the user community before doing anything more radical.

DustyShoe committed:
Merge branch 'main' into Feature/Add-Text-tool-to-canvas

lstein committed:
Merge branch 'main' into Feature/Add-Text-tool-to-canvas
